tests/unit/ops/test_fused_linear_cross_entropy.py

```
@@ -0,0 +1,108 @@
+    early_cast = saved_gradient * token_weights.to(torch.bfloat16)[:, None]
+    torch.testing.assert_close(hidden.grad, expected, rtol=0, atol=0)
+    assert not torch.equal(expected, early_cast)
+
+
+def test_rejects_trainable_head_and_wrong_target_dtype():
+    hidden = torch.randn(2, 3, requires_grad=True)
+    weight = torch.randn(5, 3, requires_grad=True)
+    target = torch.tensor([1, 2])
+    with pytest.raises(ValueError, match="frozen LM head"):
+        fused_ce.frozen_linear_cross_entropy(hidden, weight, target)
+    with pytest.raises(ValueError, match="torch.long"):
```

Ruff (0.15.21)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)